### DISCUSSION TITLE
How to use urql without backend / optimistic response mutation persistence?

### DISCUSSION THREAD

**Discussion by MStumpp** [urql-graphql/urql#2186] [Q&A] [answered]
According to [GraphCache Offline](https://formidable.com/open-source/urql/docs/graphcache/offline), urql supports a mechanism to handle offline situations.

In addition, I am looking for a mechanism to handle server outages. In other words, I want my React app to be fully functional using local cache only including persistence via using IndexedDB/localStorage (in order to keep data over browser refreshes).

My current setup uses offlineExchange plugin with optimistic responses as well as updates definitions.

```
const cache = offlineExchange({
  schema,
  storage,
  updates: {
    /* ... */
  },
  optimistic: {
    /* ... */
  },
});

const client = createClient({
  url: SERVER_GRAPHQL_ENDPOINT,
  exchanges: [dedupExchange, cache, fetchExchange],
  fetchOptions: () => {
    return {
      headers: { authorization: `Bearer ${accessToken}` },
    };
  },
});
```

However, the app works as expected with the server up and running. When shutting the server down, mutations aren't executed as expected. I can see processing of optimistic responses but since the backend call results in an error, ui changes get reversed.

**Comment by kitten** [comment, marked as answer]
> but since the backend call results in an error

Currently, we only react to these kinds of errors when the API is _actually_ offline, meaning, you'll see that `offlineExchange` has a condition that only reacts when the client is offline: https://github.com/FormidableLabs/urql/blob/0b1c3d792b4bd2ca8699f77e06d05aca026926eb/exchanges/graphcache/src/offlineExchange.ts#L60-L67

Basically, there's no standard way for us to react when the server instead is offline, as we wouldn't expect this condition. It's possible we could extend this to more conditions though, i.e. `!!error.networkError`

**Reply by MStumpp** [reply]
@kitten awesome. Basically the existing implementation is sufficient because `error.networkError.message` contains `failed to fetch` in server offline case.

However, I found one (edge) case `offlineExchange` doesn't work: Assuming the web app is fetched for the first time (offline cache is empty) and GraphQl server is offline, queries and mutations don't work as expected.

In that case, I found out that optimistic response as well as update functions do get called 

```
offlineExchange({ 
  /* ... */
  updates: {
   /* ... */
   Mutation: {
       insert_object(_result, _args, cache, _info) {
          const fields = cache.inspectFields('query_root');
          console.log(fields); // => []
       }
  },
  optimistic: {
    /* ... */
  },
})
```

but cache.inspectFields('query_root') is empty.

Urql seems to require a query to be called at least once to get "registered" in the client. After that, Urql works as expected in the offline scenario.

Is it possible to get Urql also work in the case when the Graphql server is offline at the beginning?
